### PR TITLE
Support dedicated load job project ID in BigQuery connector (#171)

### PR DESCRIPTION
**Pull request by digger**
Allow a BQ load to a table under a specific Project ID use a different Project ID to run the load job.

### CODE REVIEW

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/BigQueryOutputConfiguration.java [hunk 1]
```diff
@@ -252,24 +252,49 @@ public static boolean getCleanupTemporaryDataFlag(Configuration conf) {
...
+   *
+   * If the {@link BigQueryConfiguration#OUTPUT_PROJECT_ID_KEY} is missing, this
+   * resolves to referencing the {@link BigQueryConfiguration#PROJECT_ID_KEY} key.
    *
    * @param conf the configuration to reference the keys from.
    * @return the project id based on the given configuration.
    * @throws IOException if a required key is missing.
    */
-  public static String getProjectId(Configuration conf) throws IOException {
+  public static String getOutputProjectId(Configuration conf) throws IOException {
```
Comment (functicons) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/BigQueryOutputConfiguration.java]: This is a public method, to keep backwards compatible, I don't think we should rename it. Also, in the context of BigQueryOutputConfiguration, getProjectId is better than getOutputProjectId. But we can add comments to explain the difference between the 2 project IDs.

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/BigQueryOutputConfiguration.java [hunk 2]
```diff
@@ -252,24 +252,49 @@ public static boolean getCleanupTemporaryDataFlag(Configuration conf) {
...
+   * Gets the project id to be used to run BQ load job based on the given configuration.
+   *
+   * If the {@link BigQueryConfiguration#PROJECT_ID_KEY} is missing, this resolves to
+   * referencing the {@link BigQueryConfiguration#OUTPUT_PROJECT_ID_KEY} key.
+   *
+   * @param conf the configuration to reference the keys from.
+   * @return the project id based on the given configuration.
+   * @throws IOException if a required key is missing.
+   */
+  public static String getJobProjectId(Configuration conf) throws IOException {
```
Comment (functicons) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/BigQueryOutputConfiguration.java]: Add comments about the difference between output project ID and job project ID.

**Comment by digger** [comment]
Thanks for the feedback. The code is updated to address the comments.

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/BigQueryOutputConfiguration.java [hunk 3]
```diff
@@ -234,7 +234,7 @@ public static void validateConfiguration(Configuration conf) throws IOException
     ConfigurationUtil.getMandatoryConfig(conf, REQUIRED_KEYS);
 
     // Run through the individual getters as they manage error handling.
-    getProjectId(conf);
+    getJobProjectId(conf);
```
Comment (functicons) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/BigQueryOutputConfiguration.java]: Do we want to check both getProjectId() and getJobProjectId() here?

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/FederatedBigQueryOutputCommitter.java [hunk 4]
```diff
@@ -59,7 +59,7 @@ public void commitJob(JobContext context) throws IOException {
     // Get the destination configuration information.
     Configuration conf = context.getConfiguration();
     TableReference destTable = BigQueryOutputConfiguration.getTableReference(conf);
-    String destProjectId = BigQueryOutputConfiguration.getProjectId(conf);
+    String destProjectId = BigQueryOutputConfiguration.getJobProjectId(conf);
```
Comment (functicons) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/FederatedBigQueryOutputCommitter.java]: Rename it to jobProjectId?

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/IndirectBigQueryOutputCommitter.java [hunk 5]
```diff
@@ -59,7 +59,7 @@ public void commitJob(JobContext context) throws IOException {
     // Get the destination configuration information.
     Configuration conf = context.getConfiguration();
     TableReference destTable = BigQueryOutputConfiguration.getTableReference(conf);
-    String destProjectId = BigQueryOutputConfiguration.getProjectId(conf);
+    String destProjectId = BigQueryOutputConfiguration.getJobProjectId(conf);
```
Comment (functicons) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/output/IndirectBigQueryOutputCommitter.java]: ditto.

**Comment by digger** [comment]
Thanks, I agree with the comments, the change is updated.

**Pull request merged**